Finally, we can filter the dataframe to only include the Falcon 9.

# Data Cleaning & Wrangling


  In this section, we are going to clean and wrangle our collected data. I initially planned to split this section into two parts, but sometimes it's difficult to draw a clear boundary between these two processes. Therefore, I decided to cover them together.

  During data cleaning, our main focus is to handle missing values and outliers. Personally, I prefer identifying outliers first and then dealing with missing values. This is because the technique used to handle outliers can sometimes depend on how many outliers are present.

  To detect outliers, we can use summary statistics to compute the interquartile range (IQR) and apply the outlier rule to flag unusual values in each column. Once we identify potential outliers, we can assess whether the values are truly outliers or just appear extreme. If confirmed, a common approach is to convert these outliers into missing values, so we can treat them alongside the original missing values.

  For handling missing values, there are several common techniques. At the university level, methods like mean, median, and mode imputation are widely taught. If the data is categorical, the mode is usually the best choice. For numerical data, we need to check the distribution. If the data is clearly skewed (i.e., not normally distributed), the median is often more appropriate; otherwise, the mean can be used. In such cases, creating a histogram can help determine the skewness and guide the choice of imputation method.

  Initially, I found it confusing why we were visualizing data at this stage since visualization is typically emphasized during feature engineering. However, it makes sense to use histograms here to guide missing value treatment.

  While working on this project, I also explored several new methods for imputing missing values. One particularly interesting technique involves using machine learning models like K-Nearest Neighbors (KNN) or regression to fill in missing values by leveraging correlations within the data. This way was not usually taught in the University level courses.


  Missing data can be categorized into three types:

  1. **Missing Completely at Random (MCAR)**
    - For example, someone forgets to fill in their age on a survey without any connection to other columns.

  2. **Missing at Random (MAR)**
    - For example, people with lower education levels might intentionally or unintentionally skip the salary question.

  3. **Missing Not at Random (MNAR)**
    - For example, a woman may choose not to answer a question about her weight, making the missingness related to the value itself.

  MCAR can typically be handled using simple techniques such as mean, median, or mode imputation.

  MAR, on the other hand, can be addressed using regression or K-Nearest Neighbors (K-NN), since the missing values are related to other observed variables. For MAR, it's appropriate to use values from related columns to predict the missing ones.

  MNAR is usually problematic for most well-known imputation methods.

  At first, I was concerned that using regression for MAR might lead to multicollinearity issues because the method relies on correlated features. However, it's important to note that the missing value is the **response**, not a **predictor**, so multicollinearity is not a concern here.That said, if we later use these imputed variables as **features** in a predictive model, multicollinearity could become a problem. But that would be a separate issue addressed during feature selection or model building.

## Filter data by Falcon 9
Typically, data filtering is considered part of the data wrangling stage and is done after the cleaning process. However, I chose to filter the data for Falcon 9 first for three reasons:

1. Efficiency – I wanted to avoid having to clean the values separately for Falcon 9 and non–Falcon 9 launches.

2. Project scope – We already know that the final analysis will only involve Falcon 9 data.

3. Consistency with reference work – The original IBM team also performed this filtering step early in their process.

For these reasons, I decided to filter the data before the full cleaning and wrangling process.



In [22]:
# This is actually the wrangling process where we already know what type of the data that we are interested into.
data_falcon9 = data[data['BoosterVersion'] == 'Falcon 9'].reset_index(drop=True)
data_falcon9

FlightNumber       Date BoosterVersion  PayloadMass Orbit    LaunchSite  \
0              6 2010-06-04       Falcon 9          NaN   LEO  CCSFS SLC 40   
1              8 2012-05-22       Falcon 9        525.0   LEO  CCSFS SLC 40   
2             10 2013-03-01       Falcon 9        677.0   ISS  CCSFS SLC 40   
3             11 2013-09-29       Falcon 9        500.0    PO   VAFB SLC 4E   
4             12 2013-12-03       Falcon 9       3170.0   GTO  CCSFS SLC 40   
..           ...        ...            ...          ...   ...           ...   
85           102 2020-09-03       Falcon 9      15600.0  VLEO    KSC LC 39A   
86           103 2020-10-06       Falcon 9      15600.0  VLEO    KSC LC 39A   
87           104 2020-10-18       Falcon 9      15600.0  VLEO    KSC LC 39A   
88           105 2020-10-24       Falcon 9      15600.0  VLEO  CCSFS SLC 40   
89           106 2020-11-05       Falcon 9       3681.0   MEO  CCSFS SLC 40   

        Outcome  Flights  GridFins  Reused   Legs                LandingPad  \
0     None None        1     False   False  False                      None   
1     None None        1     False   False  False                      None   
2     None None        1     False   False  False                      None   
3   False Ocean        1     False   False  False                      None   
4     None None        1     False   False  False                      None   
..          ...      ...       ...     ...    ...                       ...   
85    True ASDS        2      True    True   True  5e9e3032383ecb6bb234e7ca   
86    True ASDS        3      True    True   True  5e9e3032383ecb6bb234e7ca   
87    True ASDS        6      True    True   True  5e9e3032383ecb6bb234e7ca   
88    True ASDS        3      True    True   True  5e9e3033383ecbb9e534e7cc   
89    True ASDS        1      True   False   True  5e9e3032383ecb6bb234e7ca   

    Block  ReusedCount Serial   Longitude   Latitude  
0     1.0            0  B0003  -80.577366  28.561857  
1     1.0            0  B0005  -80.577366  28.561857  
2     1.0            0  B0007  -80.577366  28.561857  
3     1.0            0  B1003 -120.610829  34.632093  
4     1.0            0  B1004  -80.577366  28.561857  
..    ...          ...    ...         ...        ...  
85    5.0           12  B1060  -80.603956  28.608058  
86    5.0           13  B1058  -80.603956  28.608058  
87    5.0           12  B1051  -80.603956  28.608058  
88    5.0           12  B1060  -80.577366  28.561857  
89    5.0            8  B1062  -80.577366  28.561857  

[90 rows x 17 columns]


Now that we have removed some values we should reset the FlgihtNumber column

In [23]:
data_falcon9.loc[:,'FlightNumber'] = list(range(1, data_falcon9.shape[0]+1))
data_falcon9

FlightNumber       Date BoosterVersion  PayloadMass Orbit    LaunchSite  \
0              1 2010-06-04       Falcon 9          NaN   LEO  CCSFS SLC 40   
1              2 2012-05-22       Falcon 9        525.0   LEO  CCSFS SLC 40   
2              3 2013-03-01       Falcon 9        677.0   ISS  CCSFS SLC 40   
3              4 2013-09-29       Falcon 9        500.0    PO   VAFB SLC 4E   
4              5 2013-12-03       Falcon 9       3170.0   GTO  CCSFS SLC 40   
..           ...        ...            ...          ...   ...           ...   
85            86 2020-09-03       Falcon 9      15600.0  VLEO    KSC LC 39A   
86            87 2020-10-06       Falcon 9      15600.0  VLEO    KSC LC 39A   
87            88 2020-10-18       Falcon 9      15600.0  VLEO    KSC LC 39A   
88            89 2020-10-24       Falcon 9      15600.0  VLEO  CCSFS SLC 40   
89            90 2020-11-05       Falcon 9       3681.0   MEO  CCSFS SLC 40   

        Outcome  Flights  GridFins  Reused   Legs                LandingPad  \
0     None None        1     False   False  False                      None   
1     None None        1     False   False  False                      None   
2     None None        1     False   False  False                      None   
3   False Ocean        1     False   False  False                      None   
4     None None        1     False   False  False                      None   
..          ...      ...       ...     ...    ...                       ...   
85    True ASDS        2      True    True   True  5e9e3032383ecb6bb234e7ca   
86    True ASDS        3      True    True   True  5e9e3032383ecb6bb234e7ca   
87    True ASDS        6      True    True   True  5e9e3032383ecb6bb234e7ca   
88    True ASDS        3      True    True   True  5e9e3033383ecbb9e534e7cc   
89    True ASDS        1      True   False   True  5e9e3032383ecb6bb234e7ca   

    Block  ReusedCount Serial   Longitude   Latitude  
0     1.0            0  B0003  -80.577366  28.561857  
1     1.0            0  B0005  -80.577366  28.561857  
2     1.0            0  B0007  -80.577366  28.561857  
3     1.0            0  B1003 -120.610829  34.632093  
4     1.0            0  B1004  -80.577366  28.561857  
..    ...          ...    ...         ...        ...  
85    5.0           12  B1060  -80.603956  28.608058  
86    5.0           13  B1058  -80.603956  28.608058  
87    5.0           12  B1051  -80.603956  28.608058  
88    5.0           12  B1060  -80.577366  28.561857  
89    5.0            8  B1062  -80.577366  28.561857  

[90 rows x 17 columns]

## Fill the null
In the original IBM dataset, outliers were not addressed initially. **This could potentially be an issue** and may be worth revisiting later to see if handling outliers improves model performance. For now, however, we are only focusing on filling missing values using the mean.

In [24]:
data_falcon9.isnull().sum()

FlightNumber       0
Date               0
BoosterVersion     0
PayloadMass        5
Orbit              0
LaunchSite         0
Outcome            0
Flights            0
GridFins           0
Reused             0
Legs               0
LandingPad        26
Block              0
ReusedCount        0
Serial             0
Longitude          0
Latitude           0
dtype: int64

Here, I want to further discuss how we detect missing values and decide how to handle them. For example, seeing the number 5 in the `PayloadMass` column isn't very informative by itself—it could mean 5 out of 10 values are missing or 5 out of 1,000. This makes a big difference. When a dataset has a significant amount of missing data, it's often recommended to drop the column entirely. A commonly used guideline is as follows:

1. **0%–10% missing values**: Can be imputed using the mean, median, or mode.
2. **10%–30% missing values**: Consider using regression, K-Nearest Neighbors (KNN), or Multiple Imputation by Chained Equations (MICE).
3. **Over 30% missing values**: It is generally advised to drop the column, as the missing data may be too extensive for reliable statistical inference.

This is a different criterion compared to assessing whether the data is MCAR, MAR, or MNAR.


In [25]:
data_falcon9.isnull().sum()/len(data_falcon9)*100

FlightNumber       0.000000
Date               0.000000
BoosterVersion     0.000000
PayloadMass        5.555556
Orbit              0.000000
LaunchSite         0.000000
Outcome            0.000000
Flights            0.000000
GridFins           0.000000
Reused             0.000000
Legs               0.000000
LandingPad        28.888889
Block              0.000000
ReusedCount        0.000000
Serial             0.000000
Longitude          0.000000
Latitude           0.000000
dtype: float64

Fortunately, `PayloadMass` has only about **5.6%** missing values, so replacing them with the mean or median should be sufficient.

In [26]:
# This is a cleaning process.
mean_payload = data_falcon9['PayloadMass'].mean()

data_falcon9['PayloadMass'] = data_falcon9['PayloadMass'].replace(np.nan, mean_payload)

data_falcon9.head(10)

FlightNumber       Date BoosterVersion  PayloadMass Orbit    LaunchSite  \
0             1 2010-06-04       Falcon 9  6123.547647   LEO  CCSFS SLC 40   
1             2 2012-05-22       Falcon 9   525.000000   LEO  CCSFS SLC 40   
2             3 2013-03-01       Falcon 9   677.000000   ISS  CCSFS SLC 40   
3             4 2013-09-29       Falcon 9   500.000000    PO   VAFB SLC 4E   
4             5 2013-12-03       Falcon 9  3170.000000   GTO  CCSFS SLC 40   
5             6 2014-01-06       Falcon 9  3325.000000   GTO  CCSFS SLC 40   
6             7 2014-04-18       Falcon 9  2296.000000   ISS  CCSFS SLC 40   
7             8 2014-07-14       Falcon 9  1316.000000   LEO  CCSFS SLC 40   
8             9 2014-08-05       Falcon 9  4535.000000   GTO  CCSFS SLC 40   
9            10 2014-09-07       Falcon 9  4428.000000   GTO  CCSFS SLC 40   

       Outcome  Flights  GridFins  Reused   Legs LandingPad  Block  \
0    None None        1     False   False  False       None    1.0   
1    None None        1     False   False  False       None    1.0   
2    None None        1     False   False  False       None    1.0   
3  False Ocean        1     False   False  False       None    1.0   
4    None None        1     False   False  False       None    1.0   
5    None None        1     False   False  False       None    1.0   
6   True Ocean        1     False   False   True       None    1.0   
7   True Ocean        1     False   False   True       None    1.0   
8    None None        1     False   False  False       None    1.0   
9    None None        1     False   False  False       None    1.0   

   ReusedCount Serial   Longitude   Latitude  
0            0  B0003  -80.577366  28.561857  
1            0  B0005  -80.577366  28.561857  
2            0  B0007  -80.577366  28.561857  
3            0  B1003 -120.610829  34.632093  
4            0  B1004  -80.577366  28.561857  
5            0  B1005  -80.577366  28.561857  
6            0  B1006  -80.577366  28.561857  
7            0  B1007  -80.577366  28.561857  
8            0  B1008  -80.577366  28.561857  
9            0  B1011  -80.577366  28.561857

In [27]:
data_falcon9.isnull().sum()/len(data_falcon9)*100

FlightNumber       0.000000
Date               0.000000
BoosterVersion     0.000000
PayloadMass        0.000000
Orbit              0.000000
LaunchSite         0.000000
Outcome            0.000000
Flights            0.000000
GridFins           0.000000
Reused             0.000000
Legs               0.000000
LandingPad        28.888889
Block              0.000000
ReusedCount        0.000000
Serial             0.000000
Longitude          0.000000
Latitude           0.000000
dtype: float64

In [28]:
data_falcon9.dtypes

FlightNumber               int64
Date              datetime64[ns]
BoosterVersion            object
PayloadMass              float64
Orbit                     object
LaunchSite                object
Outcome                   object
Flights                    int64
GridFins                    bool
Reused                      bool
Legs                        bool
LandingPad                object
Block                    float64
ReusedCount                int64
Serial                    object
Longitude                float64
Latitude                 float64
dtype: object

We are not going to deal with the **LandingPad** column. There are a couple of reasons for this. The main reason is that the `None` values can be considered a valid case—during the early stages of SpaceX, they may not have planned to recover the first-stage boosters, and therefore did not assign a landing pad. Another reason is that, if we treat these as missing values, the percentage is already around 28%, which may be too high for imputation. We might even drop this column later. However, I believe the primary reason is the first one.


Since we have completed most of the data cleaning and wrangling, the last step is to ensure that we have properly cleaned the target variable (i.e., the response variable). Let's take a look at our outcome.

In [29]:
landing_outcomes = data_falcon9['Outcome'].value_counts()
landing_outcomes


Outcome
True ASDS      41
None None      19
True RTLS      14
False ASDS      6
True Ocean      5
False Ocean     2
None ASDS       2
False RTLS      1
Name: count, dtype: int64

It is not difficult to see that this outcome column can actually be split into two separate columns: whether the landing was successful or not, and the method of landing (e.g., ocean). Let's now count the number of landing methods that resulted in failure.

In [30]:
bad_outcomes=set()
for i,outcome in enumerate(landing_outcomes.keys()):
    if outcome[:4] == "True":
      continue
    else:
      bad_outcomes.add(outcome)
bad_outcomes

{'False ASDS', 'False Ocean', 'False RTLS', 'None ASDS', 'None None'}

Now, we are going to create a new column to store values where 1 represents a successful landing and 0 represents a failure, based on the outcome and the failed landing methods.

In [31]:
data_falcon9['Class'] =  (~data_falcon9['Outcome'].isin(bad_outcomes)).astype(int)
data_falcon9.head(10)

FlightNumber       Date BoosterVersion  PayloadMass Orbit    LaunchSite  \
0             1 2010-06-04       Falcon 9  6123.547647   LEO  CCSFS SLC 40   
1             2 2012-05-22       Falcon 9   525.000000   LEO  CCSFS SLC 40   
2             3 2013-03-01       Falcon 9   677.000000   ISS  CCSFS SLC 40   
3             4 2013-09-29       Falcon 9   500.000000    PO   VAFB SLC 4E   
4             5 2013-12-03       Falcon 9  3170.000000   GTO  CCSFS SLC 40   
5             6 2014-01-06       Falcon 9  3325.000000   GTO  CCSFS SLC 40   
6             7 2014-04-18       Falcon 9  2296.000000   ISS  CCSFS SLC 40   
7             8 2014-07-14       Falcon 9  1316.000000   LEO  CCSFS SLC 40   
8             9 2014-08-05       Falcon 9  4535.000000   GTO  CCSFS SLC 40   
9            10 2014-09-07       Falcon 9  4428.000000   GTO  CCSFS SLC 40   

       Outcome  Flights  GridFins  Reused   Legs LandingPad  Block  \
0    None None        1     False   False  False       None    1.0   
1    None None        1     False   False  False       None    1.0   
2    None None        1     False   False  False       None    1.0   
3  False Ocean        1     False   False  False       None    1.0   
4    None None        1     False   False  False       None    1.0   
5    None None        1     False   False  False       None    1.0   
6   True Ocean        1     False   False   True       None    1.0   
7   True Ocean        1     False   False   True       None    1.0   
8    None None        1     False   False  False       None    1.0   
9    None None        1     False   False  False       None    1.0   

   ReusedCount Serial   Longitude   Latitude  Class  
0            0  B0003  -80.577366  28.561857      0  
1            0  B0005  -80.577366  28.561857      0  
2            0  B0007  -80.577366  28.561857      0  
3            0  B1003 -120.610829  34.632093      0  
4            0  B1004  -80.577366  28.561857      0  
5            0  B1005  -80.577366  28.561857      0  
6            0  B1006  -80.577366  28.561857      1  
7            0  B1007  -80.577366  28.561857      1  
8            0  B1008  -80.577366  28.561857      0  
9            0  B1011  -80.577366  28.561857      0

Let's quickly look at the landing success rate based on the current data.


In [32]:
# what is the difference between data_falcon9['Class'] and data_falcon9[['Class']]
data_falcon9['Class'].mean()

# Note: the class is not balance. We may want to watch this.

np.float64(0.6666666666666666)

Although the class distribution in our dataset is not extremely imbalanced (approximately 67% success and 33% failure), it is still important to be aware of this imbalance when training and evaluating machine learning models. A model trained on such data may become biased toward the majority class, leading to misleadingly high accuracy while underperforming on the minority class. To mitigate this, we will evaluate the model using metrics such as precision and recall to gain a more accurate understanding of its performance across both classes

# Acknowledgments

This project was originally conceptualized and implemented by the following contributors:

- <a href="https://www.linkedin.com/in/yan-luo-96288783/">Yan Luo</a>  
- <a href="https://www.linkedin.com/in/nayefaboutayoun/">Nayef Abou Tayoun</a>  
- IBM Corporation
